src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+
+#ifdef USE_ALT_HOLD_MODE
+
+#include "math.h"
+#include "build/debug.h"
+
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "fc/rc.h"
+#include "flight/position.h"
+#include "flight/failsafe.h"
+#include "sensors/acceleration.h"
+#include "rx/rx.h"
+
+#include "alt_hold.h"
+
+typedef struct {
+    float kp;
+    float kd;
+    float ki;
+    float lastAltitude;
+    float integral;
+} simplePid_t;
+
+simplePid_t simplePid;
+
+altHoldState_t altHoldState;
+
+#define ALT_HOLD_PID_P_GAIN  0.01f
+#define ALT_HOLD_PID_I_GAIN  0.003f
+#define ALT_HOLD_PID_D_GAIN  0.01f
+
+static pt2Filter_t altHoldDeltaLpf;
+static const float taskIntervalSeconds = 1.0f / ALTHOLD_TASK_RATE_HZ; // i.e. 0.01 s
+
+float altitudePidCalculate(void)
+{
+    const float altErrorCm = altHoldState.targetAltitudeCm - altHoldState.measuredAltitudeCm;
+
+    // P
+    const float pOut = simplePid.kp * altErrorCm;
+
+    // I
+    simplePid.integral += altErrorCm * taskIntervalSeconds; // cm * seconds
+    float iOut = simplePid.ki * simplePid.integral;
+    // arbitrary limit on iTerm, same as for gps_rescue, +/-20% throttle
+    iOut = constrainf(iOut, -200.0f, 200.0f); 
+
+    // D
+    const float derivative = (simplePid.lastAltitude - altHoldState.measuredAltitudeCm) / taskIntervalSeconds; // cm/s
+    simplePid.lastAltitude = altHoldState.measuredAltitudeCm;
+    float dOut = simplePid.kd * derivative;
+    // PT2 filtering at altitude_d_lpf / 100 cutoff, default 1s, the same as for GPS Rescue 
+    dOut = pt2FilterApply(&altHoldDeltaLpf, dOut);
+
+    // F
+    const float fOut = altholdConfig()->alt_hold_pid_d * altHoldState.targetAltitudeDelta;
+    // if error is used, we get a 'free kick' in derivative from changes in the target value
+    // but this is delayed by the smoothing, leading to lag and overshoot.
+    // calculating feedforward separately avoids the filter lag.
+
+    const float output = pOut + iOut + dOut + fOut;
+
+    DEBUG_SET(DEBUG_ALTHOLD, 4, lrintf(pOut));
+    DEBUG_SET(DEBUG_ALTHOLD, 5, lrintf(iOut));
+    DEBUG_SET(DEBUG_ALTHOLD, 6, lrintf(dOut));
+    DEBUG_SET(DEBUG_ALTHOLD, 7, lrintf(fOut));
```

Comment:
Note slot 3 is not used here (used for total output before).

Reply:
Yes, it's free for some other thing that might arise.

Reply:
Thanks, updated the configurator PR

- [ ] https://github.com/betaflight/betaflight-configurator/pull/4130

Reply:
Thanks @haslinghuis